File: 2020/01/ruby/solution.rb
#!/usr/bin/env ruby
require "set"

class ReportRepair
  def initialize(input)
    @entries = input.split.map(&:to_i)
  end

  def find_2020
    a, b = find_pair(2020)
    a * b
  end

  def find_pair(sum)
    # since the entries are bounded we can use them as members in a set
    seen = Set.new

    @entries.each do |e|
      other = sum - e
      return [e, other] if seen.include?(other)

      seen << e
    end
    raise "No matching pair found"
  end

  def find_triplet(sum)
    # brute force for a change
    @entries.each do |a|
      @entries.each do |b|
        @entries.each do |c|
          return [a, b, c] if a + b + c == sum
        end
      end
    end
    raise "No matching triplet found"
  end
end

if $PROGRAM_NAME == __FILE__
  rr = ReportRepair.new(File.read("input.txt"))
  puts "Part 1"
  puts rr.find_2020

  puts "Part 2"
  a, b, c = rr.find_triplet(2020)
  puts a * b * c
end

File: 2020/01/python/solution.py
#!/usr/bin/env python3
class ReportRepair(object):
    def __init__(self, input):
        self.entries = [int(line) for line in input.split()]

    def find_pair(self, sum):
        seen = set()

        for e in self.entries:
            other = sum - e
            if other in seen:
                return (other, e)
            seen.add(e)

        raise "No matching pair found"

    def find_triplet(self, sum):
        for a in self.entries:
            for b in self.entries:
                for c in self.entries:
                    if a + b + c == sum:
                        return (a, b, c)

        raise "No matching triplet found"


if __name__ == '__main__':
    with open("input.txt") as f:
        input = f.read()
    rr = ReportRepair(input)
    # print(repr(rr.entries))

    print("Part 1")
    a, b = rr.find_pair(2020)
    print(a * b)

    print("Part 2")
    a, b, c = rr.find_triplet(2020)
    print(a * b * c)
